### ISSUE TITLE
cypher statement failing - Unknown identifier `  INTERNAL_AGGREGATE8d07c913

### ISSUE THREAD

**Issue by Infiniverse** [neo4j/neo4j#8485]
The attached statement does not run on Neo4j 2.3.x (tested on 2.3.8), and instead issues warnings:

    WARNING: Unknown identifier `  INTERNAL_AGGREGATE8d07c913-1eb3-4574-b62b-b2a9dddbd42c`.
    Unknown identifier `  INTERNAL_AGGREGATE4ea6739e-f25a-4f19-b099-60519ff03cd1`.
    Unknown identifier `  AGGREGATION498`.
    Unknown identifier `billUK`.
    Unknown identifier `billMobile`.
    Unknown identifier `billInternational`.
    Unknown identifier `  INTERNAL_AGGREGATEca8affd6-a1f2-47a3-b395-155727b9e713`.
    Unknown identifier `  INTERNAL_AGGREGATEb4bf602f-803d-474e-9362-580a25153846`.

The statement is:

    WITH "2016-11" AS month, "November" as monthName
    MATCH (s:Source),(p:PhoneLine)--(so:SalesOrder)--(i:Invoiceable { state: "DRAFT"})
    WHERE s.src = p.msisdn
    WITH month, monthName, so, s, i, so.phone_callingPlan = "PAYT" AS billUK, NOT     HAS(so.boltOns_mobile) AS billMobile, NOT HAS(so.boltOns_international) AS billInternational
    WITH month, monthName, so, s, i, billUK, billMobile, billInternational
    MATCH (s)--(c:Call)--(ch:Charge)--(t:Tariff)
    WHERE c.start STARTS WITH month
    WITH monthName, s, i, t.type AS type, COUNT(c) AS calls, SUM(ch.total) AS chargePence, SUM(c.billsec) AS seconds, ROUND(10 * SUM(TOFLOAT(c.billsec))/60)/10 AS minutes, CASE
    WHEN t.type = "UK Landline" AND NOT billUK THEN FALSE
    WHEN t.type = "UK Mobile" AND NOT billMobile THEN FALSE
    WHEN t.type = "International" AND NOT billInternational THEN FALSE
    ELSE true
    END AS chargeable
    WITH monthName, s, i, type, calls, CASE chargeable WHEN TRUE THEN chargePence ELSE 0 END AS chargePence, seconds, minutes
    WITH monthName, type, i, "0" + RIGHT(s.src, SIZE(s.src) - 2) + ": " + calls + " calls, lasting " + minutes + " minutes." as desc, chargePence / 100 AS charge
    CREATE (i)-[:HAS_LINEITEM]->(item:LineItem:_LineItem { title: "Phone Calls: " + type + " (" + monthName + ")" })
    SET item.id = 101
    SET item.price = charge
    SET item.description = desc
    SET item.frequency = "MONTHLY"
    SET item.type = "TELEPHONY"
    RETURN item;

It does issues the same warning when run with PROFILE, and emits no profile information.

When I remove the second WITH:

        WITH month, monthName, so, s, i, so.phone_callingPlan = "PAYT" AS billUK, NOT     HAS(so.boltOns_mobile) AS billMobile, NOT HAS(so.boltOns_international) AS billInternational
        WITH month, monthName, so, s, i, billUK, billMobile, billInternational

The statement runs fine.

Is this a bug?

**Comment by Mats-SX** [comment]
Hello @Infiniverse and thanks for reporting this issue (it is indeed a bug).

I believe that this has been fixed in the `3.x` timeline. Is it possible for you to either share a small setup dataset that enables us to reproduce (just running the query on an empty `2.3.8` database doesn't fail), or to try it yourself on `3.0` and/or `3.1`, to see if the bug is present there as well?

**Comment by Infiniverse** [comment]
Hi Mats,

I’m don’t think that we currently have the resource to put the dataset together - the data in question is live customer data, and so would need to be anonymised first which I think might take some time :/.

Joe

**Comment by Mats-SX** [comment]
@Infiniverse Is it then possible for you to try and copy the dataset into a `3.1` Neo4j instance and run the query? I believe that an upgrade will fix this query issue for you.

**Comment by Infiniverse** [comment]
I'll give that a go. Unfortunately I'm not in a position to upgrade against 3.1 as we have other dependencies which have not been ported to 3.1 yet, so we do need a fix in 2.3.x.

Strangely the query that ran last month is no longer running this month with the duplicate WITH removed. This means that we can't run this query at all right now :/ :(

**Comment by Mats-SX** [comment]
@Infiniverse I see. I did a quick investigation now and did manage to reproduce the error. Will try and submit a fix for it that would ship with `2.3.10`.

**Comment by Infiniverse** [comment]
Excellent. Is it already committed to a prerelease branch so I could try it?

**Comment by Infiniverse** [comment]
It doesn't look like it's on the 2.3 main branch yet. Is it pushed to another repository? (I'd like to test please! :) )

**Comment by Mats-SX** [comment]
@Infiniverse No, it isn't committed yet.

**Comment by Infiniverse** [comment]
Any chance that you could? I'm desperate to test it :).
(We're still running java 7 here, and so even getting neo4j 3.1 up and running is proving to be a challenge).

**Comment by Mats-SX** [comment]
No, I can not. I haven't been able to find or fix the issue yet.

**Comment by Infiniverse** [comment]
Oh, ok. I thought that you were saying that you'd seen the problem on 2.x and were working on a fix:

> I did a quick investigation now and did manage to reproduce the error. Will try and submit a fix for it that would ship with 2.3.10.

In that case I'll load the database into 3.1 today and check that the problem has been resolved there.

What's strange about the problem is that it appears to be data related. That is, the query has run in the past, but later (when there are more nodes in the graph) it stops working. I'm confused by that - is it the query optimiser that's breaking things?

**Comment by Mats-SX** [comment]
@Infiniverse You are correct. I have managed to reproduce the problem on `2.3.9`. What I have not done is find the _cause_ of the problem.

Has the query run in the past such that it produced nonempty results?

**Comment by Infiniverse** [comment]
> @Infiniverse <https://github.com/Infiniverse> You are correct. I have managed to reproduce the problem on 2.3.9. What I have not done is find the cause of the problem.
> Has the query run in the past such that it produced nonempty results?
> 


Yes, it’s been running reliably for 6 months or so (once a month).
However, periodically it’s failed with the INTERNAL_AGGREGATE errors.
It’s now not running at all and always failing.
:(

**Comment by Infiniverse** [comment]
Ok - I can confirm that it runs fine on Neo4j (enterprise) 3.0.8.

**Comment by Mats-SX** [comment]
@Infiniverse I've found the bug, but it's not easy to fix. The bug is deep inside the rule planner, which was deprecated in `3.0` and will be removed from our upcoming `3.2` release. The general recommendation is, again, to upgrade.

However, I did find a workaround for your specific query. If you add the three `bill*` variables to the `WITH` clause that does all the aggregations, your query should execute fine:

```
WITH "2016-11" AS month, "November" AS monthName
MATCH (s:Source), (p:PhoneLine)--(so:SalesOrder)--(i:Invoiceable { state: "DRAFT"})
WHERE s.src = p.msisdn
WITH month, monthName, so, s, i, 
  so.phone_callingPlan = "PAYT" AS billUK, 
  NOT exists(so.boltOns_mobile) AS billMobile, 
  NOT exists(so.boltOns_international) AS billInternational
WITH month, monthName, so, s, i, billUK, billMobile, billInternational
MATCH (s)--(c:Call)--(ch:Charge)--(t:Tariff)
WHERE c.start STARTS WITH month
WITH monthName, s, i, t.type AS type, 
  count(c) AS calls, 
  sum(ch.total) AS chargePence, 
  sum(c.billsec) AS seconds, 
  round(10 * sum(toFloat(c.billsec)) / 60) / 10 AS minutes, 
  billUK, billMobile, billInternational, 
  CASE
    WHEN t.type = "UK Landline" AND NOT billUK THEN false
    WHEN t.type = "UK Mobile" AND NOT billMobile THEN false
    WHEN t.type = "International" AND NOT billInternational THEN false
    ELSE true
  END AS chargeable
WITH monthName, s, i, type, calls, 
  CASE chargeable 
    WHEN true THEN chargePence 
    ELSE 0 
  END AS chargePence, seconds, minutes
WITH monthName, type, i, "0" + right(s.src, size(s.src) - 2) + ": " + calls + " calls, lasting " + minutes + " minutes." AS desc, chargePence / 100 AS charge
CREATE (i)-[:HAS_LINEITEM]->(item:LineItem:_LineItem { title: "Phone Calls: " + type + " (" + monthName + ")" })
SET item.id = 101
SET item.price = charge
SET item.description = desc
SET item.frequency = "MONTHLY"
SET item.type = "TELEPHONY"
RETURN item
```

The reason the planner has a hard time is because the way it deals with aggregations, and the variable horizons that are formed as part of this. The `CASE` expression in your query depends on variables that were projected on 'the other side' of the aggregation horizons, and this is lost in the dependency management in the planner.

If your query can be modified to not contain the updating clauses (`CREATE`, `SET`), it may be planned by the `2.3` cost planner, which does not have this problem. In `3.0`, the cost planner added support for these updating clauses as well, so the problem is gone there for that reason (and similarly for `3.1`).

I'm hoping that the workaround will work for you, and that you'd be able to continue running with that until such time as you are able to upgrade to Neo4j `3.x`.

**Comment by Infiniverse** [comment]
Ah cool. Thanks for that. It's strange, but understandable. We are kinda stuck with 2.3.x for the time being pending a rewrite of our SDN dependency, so thanks for a work around - very much appreciated.

**Comment by Infiniverse** [comment]
Was this fixed yet in 2.3.x path?

**Comment by Mats-SX** [comment]
@Infiniverse No it has not, and I wouldn't hold my breath on this one. The code has been abandoned following the removal of the rule planner in `3.2`. You'll have to live with the workaround (project variables explicitly) or upgrade, I'm afraid.

**Comment by sherfert** [comment]
Given that we have a workaround thanks to Mats, and the issue is fixed in 3.0 and subsequently, I am going to close this issue now.
